### PR TITLE
implement most of the missing challenge events

### PR DESCRIPTION
**Pull request by zacchiro**
The new class is called ChallengeEvent and by default checks whether we're
currently in a challenge or not. Additionally it can be passed arguments to
declaratively request checking that we're in a challenge of a given type and/or
participating in a given role (attacker/defender).

### CODE REVIEW

**Comment by zacchiro** [comment]
WARNING: I've submitting this now to gather feedback on the general idea.
If you like it, I'll add to this PR the implementation of all the [relevant cards](https://thronesdb.com/find?q=x%3A%22play+only+during+a%22&sort=name&view=spoiler)

There seem to be 8 of them, and currently we only have Ours is the Fury in Throneteki. Implementing the other 7 with this helper should be quasi-trivial, and will serve as further validation for this idea.

What do you think?

**Comment by ystros** [comment]
Not opposed to this approach, it should help clean up some of the repetitious logic. However, I do plan on redoing how cards are played / marshaled / ambushed / etc, so this code may end moving around or evaporating entirely.

I don't quite have 100% concrete thoughts on that yet, though, so I would say move forward with it unless there are other objections.

File: server/game/cards/challengeevent.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        //   challenge of the given type
+        //
+        // - partecipatingAs (optional): make the event playable only when the
+        //   event controller is partecipating as either 'attacker' or
+        //   'defender'
+        //
+        super(owner, cardData);
+
+        this._challengeType = challengeType;
+        this._partecipatingAs = partecipatingAs;
```
Comment (ystros) [inline server/game/cards/challengeevent.js]: Spelling - `participatingAs`. Also unsure of the convention of prefixing these variables with underscore. I understand what you're going for here, but it's not something we've done elsewhere.

Comment (zacchiro) [inline server/game/cards/challengeevent.js]: Fixed spelling and dropped _.

**Comment by zacchiro** [comment]
Following @ystros and @cryogen comments, I went ahead with this idea.

This pull request now contains the implementations for most of the challenge event cards in the meta (with the exception of There Is My Claim and The Sword in the Darkness), all based on the new ChallengeEvent class.

As part of this I also introduced a new effect to (temporarily) set the max number of challenges of a given type to an arbitrary value (= 0 in my use case, but might be other for different needs).

Reviews and comments welcome.

File: server/game/cards/challengeevent.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        super(owner, cardData);
+
+        this.challengeType = challengeType;
+        this.participatingAs = participatingAs;
+    }
+
+    canPlay(player, card) {
+        if(player !== this.controller || this !== card) {
+            return false;
+        }
```
Comment (cryogen) [inline server/game/cards/challengeevent.js]: this needs to go on the base card - that's more a note for myself than to you, though :)

File: server/game/cards/challengeevent.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+        var challenge = this.game.currentChallenge;
+        if(!challenge) {
+            return false;
+        }
+        if(this.challengeType
+           && challenge.challengeType !== this.challengeType) {
+            return false;
+        }
+        if((this.participatingAs === 'attacker'
```
Comment (cryogen) [inline server/game/cards/challengeevent.js]: I find this block quite hard to read -  I can tell you've tried to wrap to 80 chars, but it's ok to go a bit longer these days, as much as 160 even.  Perhaps putting the two bracketed conditionals on their own line might make this clearer

File: server/game/cards/events/01/winteriscoming.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const ChallengeEvent = require('../../challengeevent.js');
+
+class WinterIsComing extends ChallengeEvent {
+
+    // TODO implement restriction "(Max 1 per challenge.)"
```
Comment (cryogen) [inline server/game/cards/events/01/winteriscoming.js]: This probably means bringing events under the effects engine somehow, @ystros FYI

**Comment by cryogen** [comment]
Other than a couple of minor things ive mentioned, looks good.  It's ok to go longer than 80 chars wide, too :)

**Comment by zacchiro** [comment]
On Sun, Feb 19, 2017 at 01:22:51AM -0800, Stuart Walsh wrote:
> +        if(this.challengeType
> +           && challenge.challengeType !== this.challengeType) {
> +            return false;
> +        }
> +        if((this.participatingAs === 'attacker'
> 
> I find this block quite hard to read - I can tell you've tried to wrap
> to 80 chars, but it's ok to go a bit longer these days, as much as 160
> even.  Perhaps putting the two bracketed conditionals on their own
> line might make this clearer

More than a strict 80 char limit, I generally tend to not have multiple
logical clauses on the same line, because they make *further* changes
you make to those lines harder to read in diffs.  But here it doesn't
matter, because the left hand of the && here is just verifying the thing
is defined.  And of course I'll happily adapt to the code base style :)

Fixed in the pull request update I just pushed. Ditto for the other
"complex" boolean test in the same class.

Cheers.

**Comment by zacchiro** [comment]
just added a15c1da23e1882e88ed9593b829d60f5b7307f24, which implements There Is My Claim

**Comment by cryogen** [comment]
That reads much better, I'm ok splitting conditionals, but if they're in the same parens makes sense to keep them together

File: server/game/cards/events/04/thereismyclaim.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+const _ = require('underscore');
+
+const ChallengeEvent = require('../../challengeevent.js');
+
+class WinterIsComing extends ChallengeEvent {
```
Comment (cryogen) [inline server/game/cards/events/04/thereismyclaim.js]: copy and paste :)

**Comment by zacchiro** [comment]
On Sun, Feb 19, 2017 at 10:05:43AM -0800, Stuart Walsh wrote:
> +class WinterIsComing extends ChallengeEvent {
> 
> copy and paste :)

Gah, I suck. It's interesting (and now that I looked into it fully
understandable) that it actually worked nonetheless :)

Fixed now. Sorry for forcing you through another review/accept cycle.

**Pull request merged**